In [8]:
import numpy as np
from PIL import Image

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

import torchvision
import torchvision.transforms as transforms

In [9]:
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

In [10]:
train_data = torchvision.datasets.CIFAR10(root='./data', train=True, transform=transform, download=True)
test_data = torchvision.datasets.CIFAR10(root='./data', train=False, transform=transform, download=True)

train_loader = torch.utils.data.DataLoader(train_data, batch_size=32, shuffle=True, num_workers=2)
test_loader = torch.utils.data.DataLoader(test_data, batch_size=32, shuffle=True, num_workers=2)

In [11]:
image, label = train_data[0]

In [12]:
image.size()

torch.Size([3, 32, 32])

In [13]:
class_names = ['plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck']

In [14]:
class NeuralNet(nn.Module):
    def __init__(self):
        super().__init__()
        
        self.conv1 = nn.Conv2d(3, 12, 5) # (inputSize-kernelSize)/KernelMoveStride+1 == (32-5)/1+1=28 after convolutional we get (12channelOrFeatureMap, 28pixels, 28pixels)
        self.pool = nn.MaxPool2d(2, 2) # 2*2 area create 1 pixel leave us (12, 14, 14)
        self.conv2 = nn.Conv2d(12, 24, 5) # (14-5)/1+1 we get (24, 10, 10) -> (24,5,5) -> Flatten(24*5*5)
        self.fc1 = nn.Linear(24 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        """apply above init layer onto the input"""
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = torch.flatten(x, 1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x
        

In [15]:
net = NeuralNet()
loos_founction = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

In [16]:
# for epoch in range(30):
#     print(f"Training epoch {epoch} ...")

#     running_loss = 0.0

#     for i, data in enumerate(train_loader):
#         inputs, labels = data

#         optimizer.zero_grad()

#         outputs = net(inputs)

#         loss = loos_founction(outputs, labels)
#         loss.backward()
#         optimizer.step() 

#         running_loss += loss.item()

#     print(f"Loss: {running_loss / len(train_loader):.4f}")

In [17]:
# torch.save(net.state_dict(), "torch_trained_net_model.pth")

In [18]:
net = NeuralNet()
net.load_state_dict(torch.load("torch_trained_net_model.pth"))

<All keys matched successfully>

In [19]:
correct = 0
total = 0

net.eval()

with torch.no_grad():
    for data in test_loader:
        images, labels = data
        outputs = net(images)
        _, predicted = torch.max(outputs, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

accuracy = 100 * correct / total
print(f"Accuracy: {accuracy}%")

Accuracy: 69.4%


If it has 2 different classes 69.4% is not so good, but we now have 10 classes it quit good.

In [28]:
new_transform = transforms.Compose([
    transforms.Resize((32,32)),
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

def load_image(image_path):
    image = Image.open(image_path)
    image = new_transform(image)
    image = image.unsqueeze(0)
    return image

# Note png is 4 channels, jpg is 3 channels.
image_paths = ["example3.jpg", "example2.jpg", "example.jpg"]
images = [load_image(img) for img in image_paths]

net.eval()
with torch.no_grad():
    for image in images:
        output = net(image)
        _, predicted = torch.max(output, 1)
        print(f"Prediction: {class_names[predicted.item()]}")
        

Prediction: bird
Prediction: ship
Prediction: horse
